src/platform/posix/posix_pollq_epoll.c

```diff
@@ -134,19 +141,27 @@ nni_posix_pfd_arm(nni_posix_pfd *pfd, unsigned events)
 
 	nni_mtx_lock(&pfd->mtx);
 	if (!pfd->closing) {
-		struct epoll_event ev;
+		struct epoll_event *ev;
 		pfd->events |= events;
 		events = pfd->events;
 
-		ev.events   = events | NNI_EPOLL_FLAGS;
-		ev.data.ptr = pfd;
+		if ((ev = NNI_ALLOC_STRUCT(ev)) == NULL) {
+			return (NNG_ENOMEM);
+		}
+
+		ev->events   = events | NNI_EPOLL_FLAGS;
+		ev->data.ptr = pfd;
 
-		if (epoll_ctl(pq->epfd, EPOLL_CTL_MOD, pfd->fd, &ev) != 0) {
+		if (epoll_ctl(pq->epfd, EPOLL_CTL_MOD, pfd->fd, ev) != 0) {
```

Comment:
Same as above. Could consider reading return value and freeing before we check it.

Reply:
Actually looking at this again, I realize I was using a variable on the stack.  I didn't notice that earlier.

This was to avoid extra code paths (you have to handle NNG_ENOMEM) as well as reduce pressure on the allocator (allocations are often expensive). 

I am now thinking that valgrind's analysis is wrong here.  It's casting too wide a net -- having a variable that points at uninitialized memory is fine, as long as I don't do any reads -- valgrind used to be smarter about this.

These days I tend to prefer to the CLang analyzers that I think do a *vastly* better job than valgrind -- they catch more issues than valgrind, run faster, and don't give me false positives.

At this point, I think I'm disinclined to make this change -- "fixing" it is really working around an incorrect assessment in valgrind, and it costs something to do so.

Reply:
Looking again, its a complaint that the syscall parameters are uninitialized, and valgrind can't reach across the syscall boundary.

Maybe the memset you originally proposed is best.